mesh/nurbs.cpp

```diff
@@ -60,6 +107,165 @@ KnotVector &KnotVector::operator=(const KnotVector &kv)
    return *this;
 }
 
+int KnotVector::GetSpan(real_t u) const
+{
+   int low, mid, high;
+
+   if (u == knot(NumOfControlPoints+Order))
+   {
+      mid = NumOfControlPoints-1;
+   }
+   else if (u == knot(0))
+   {
+      mid = Order;
+   }
+   else if ((u > knot(0)) && (u < knot(NumOfControlPoints+Order)))
+   {
+      low = Order;
+      high = NumOfControlPoints;
+      mid = (low + high)/2;
+      while ( (u < knot(mid)) || (u >= knot(mid+1)) )
+      {
+         if (u < knot(mid))
+         {
+            high = mid;
+         }
+         else
+         {
+            low = mid;
+         }
+         mid = (low + high)/2;
+      }
+   }
+   else
+   {
+      mfem_error("Knot location outside of the range of the KnotVector");
+   }
+
+   return mid;
+}
+
+real_t KnotVector::GetGreville(int i) const
+{
+   real_t sum = 0.0;
+   for (int j = 1; j < Order+1; j++) { sum += knot[i + j]; }
+   return sum/real_t(Order);
+}
+
+real_t KnotVector::GetBotella(int i) const
+{
+   constexpr int itermax = 10;
+   constexpr real_t tol = 1e-8;
+
+   Vector grad(Order+1);
+   Vector hess(Order+1);
+
+   real_t u,xi;
+   int iter, ks, o;
+
+   // Get initial guess
+   u = GetGreville(i);
+
+   // Check for a repeated knot -- include begin and end
+   if (knot[i + 1] == knot[i + Order])
+   {
+      return u;
+   }
+
+   for (iter = 0; iter < itermax; iter++)
+   {
+      ks = GetSpan (u);
+      xi = GetRefPoint(u, ks);
+      o = Order - (ks - i);
+
+      CalcDShape(grad, ks-Order, xi);
+      CalcD2Shape(hess, ks-Order, xi);
+
+      u -= (grad[o]/hess[o])*(knot(ks+1) - knot(ks));
+
+      if (fabs(grad[o])< tol) { break; }
+   }
+   if (iter >= itermax)
+   {
+      MFEM_WARNING("KnotVector::GetBotella not converged");
+      mfem::out<<"i = "<<i<<",iter = "<<iter<<", grad = "<< grad[o]<<endl;
+   }
+   return u;
+}
+
+void KnotVector::ComputeDemko() const
```

Comment:
I was testing out Greville/Botella/Demko points for various knotvectors. One thing I discovered is that it seems like if there is an internal knot with multiplicity `>= p+1` `ComputeDemko()` will fail, while the other routines wont. As an example 
```cpp
KnotVector kv(2, Vector({0.0,1.0,1.0,1.0,2.0}));
Vector greville(kv.GetNCP());
for (int i = 0; i < kv.GetNCP(); i++)
{
   greville[i] = kv.GetGreville(i);
}
mfem::out<<"Greville points : "; greville.Print(mfem::out, 32);


Vector botella(kv.GetNCP());
for (int i = 0; i < kv.GetNCP(); i++)
{
   botella[i] = kv.GetBotella(i);
}
mfem::out<<"Botella  points : "; botella.Print(mfem::out, 32);


Vector demko(kv.GetNCP());
for (int i = 0; i < kv.GetNCP(); i++)
{
   demko[i] = kv.GetDemko(i);
}
mfem::out<<"Demko    points : "; demko.Print(mfem::out, 32);
```

returns 
```cpp
Greville points : 0 0.5 1 1 1.5 2
Botella  points : 0 0.5 1 1 1.5 2

DenseMatrix::Invert() : singular matrix
```
B-splines for reference
<img width="594" alt="Screenshot 2024-10-14 at 10 07 42 PM" src="https://github.com/user-attachments/assets/b33b8f59-29da-4d05-8ca1-f593a5843d39">

Looks like its coming from the `GetInterpolant()` routine. Maybe this is expected and there is nothing to fix; in that case a note would be useful.

btw, I did a brute-force check on the Greville/Botella points for many different knotvectors and the values looked correct 👍